src/lib/y2partitioner/widgets/main_menu_bar.rb

```diff
@@ -0,0 +1,163 @@
+# Copyright (c) [2020] SUSE LLC
+#
+# All Rights Reserved.
+#
+# This program is free software; you can redistribute it and/or modify it
+# under the terms of version 2 of the GNU General Public License as published
+# by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+# more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, contact SUSE LLC.
+#
+# To contact SUSE LLC about this file by physical or electronic mail, you may
+# find current contact information at www.suse.com.
+
+require "yast"
+require "cwm"
+require "y2partitioner/widgets/execute_and_redraw"
+require "y2partitioner/actions/rescan_devices"
+
+module Y2Partitioner
+  module Widgets
+    # Main menu bar of the partitioner
+    class MainMenuBar < CWM::CustomWidget
+      include Yast::Logger
+      include ExecuteAndRedraw
+
+      # Constructor
+      def initialize
+        textdomain "storage"
+        self.handle_all_events = true
```

Comment:
Well, let me just explain basic idea of CWM ( why it was designed years ago ). Maybe whole idea is bad, I just want to explain it here. Idea was to have self-contained object ( YCP map with values originally ) that connects UI element like Push Button and its actions, help text, labels and stuff like this. So goal of widget is not to have bussiness logic there, but having glue between backend API and UI element. So ideal CWM widget should look like
```
class ValueAField < InputField
  def label
     _("Value A")
  end

  def init
     self.value = backend.value_a
  end

  def store
    backend.value_a = value
  end
end
```
 Sadly in some cases more actions are placed. Why I mention separate class is that now MenuBar is as CustomWidget so generic widget. If own class will be created, that final widget will be really just glue between UI representation of menu bar and backend that handle menu bar actions.

Reply:
Yes, I do understand the principle. And in some isolated cases this is actually useful. But moving back-end logic into a derived button is just wrong. A button (or a menu button's item) is a very dumb object that should be instantiated, not derived from.

As we can see in this use case, this can backfire easily, and it does: Moving functionality into a derived button makes that functionality unusable from anywhere else. And to a user it should not matter if it was that button that he clicked or a menu item or whatever else.

I am right now in the process of moving the actual functionality out of those derived button classes and into separate action classes so those actions are generally available. The remaining old button class will do nothing more than execute the action. This is how it should be in general. Everything else than "execute this action" (literally a one-liner) in a button class's event handling method is just plain misplaced there.

And then there is not much left for that derived button class; it's little more than one-liner methods to set its label and its ID; and the help text (which may be more than one line, but still it's very simplistic code). Most of that (label and ID) can (and should) literally be written in the UI term; there is zero gain in having an extra class with overwritten methods for `label` and `id` rather than just writing it in plain UI code:

```
PushButton(Id(:colorRed), _("Red"))
```

This is much more concise and much easier to read; you don't have to look it up in a separate file.

What we are missing right now is an elegant way to integrate this into CWM. I hope with my experiments here we might get a feeling how this could work; this PR / this PBI should be considered a technology carrier. If this works out well (which I hope), we can move that part out of the partitioner and into CWM itself.

I am talking about the general idea to use introspection and naming conventions like methods `handle_mywidgetid()` to handle an event from widget or menu item `mywidgetid`. Let's discuss this later.

Reply:
@shundhammer well, in fact CWM does not prevent that to have it in own instance. You can simply have something like
```
SimplePushButtonn.new(label, id, push_action)
```
and its implementation will be trivial like
```
class SimplePushButton < CWM::PushButton
  def initialize(label, id, push_action)
    self.widget_id = id
    @label = label # Note: this you need to reinstantiate button if language change
    @action = push_action
  end

  attr_reader: label

  def handle
     @action.call
  end
end
```